File: cpp/2975_maximize_square_area/maximize_square_area.cpp
#include <algorithm>
#include <cassert>
#include <tuple>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
  unordered_set<int> getEdges(vector<int>& fences, int border) {
    unordered_set<int> st;
    fences.push_back(1);
    fences.push_back(border);
    sort(fences.begin(), fences.end());
    for (int i = 0; i < fences.size(); i++) {
      for (int j = i + 1; j < fences.size(); j++) {
        st.insert(fences[j] - fences[i]);
      }
    }
    return st;
  }

 public:
  int maximizeSquareArea(int m, int n, vector<int>& hFences, vector<int>& vFences) {
    auto hEdges = getEdges(hFences, m);
    auto vEdges = getEdges(vFences, n);
    int res = 0;
    for (auto e : hEdges) {
      if (vEdges.contains(e)) {
        res = max(res, e);
      }
    }
    if (res == 0) {
      res = -1;
    } else {
      res = 1ll * res * res % 1000000007;
    }
    return res;
  }
};

int main() {
  vector<tuple<int, int, vector<int>, vector<int>, int>> tests{
      {4, 3, {2, 3}, {2}, 4},
      {6, 7, {2}, {4}, -1},
  };

  for (auto [m, n, hFences, vFences, expected] : tests) {
    assert(Solution().maximizeSquareArea(m, n, hFences, vFences) == expected);
  }
}

File: golang/2975_maximize_square_area/maximize_square_area.go
// Package main ...
package main

import (
	"sort"
	"testing"

	"github.com/stretchr/testify/assert"
)

func maximizeSquareArea(m int, n int, hFences []int, vFences []int) int {
	hEdges := getEdges(hFences, m)
	vEdges := getEdges(vFences, n)

	var res int64 = 0
	for e := range hEdges {
		if _, exists := vEdges[e]; exists {
			if int64(e) > res {
				res = int64(e)
			}
		}
	}

	if res == 0 {
		return -1
	}
	return int((res * res) % 1000000007)
}

func getEdges(fences []int, border int) map[int]bool {
	set := make(map[int]bool)
	list := make([]int, len(fences))

	copy(list, fences)
	list = append(list, 1, border)
	sort.Ints(list)

	for i := 0; i < len(list); i++ {
		for j := i + 1; j < len(list); j++ {
			set[list[j]-list[i]] = true
		}
	}

	return set
}

func main() {
	tests := []struct {
		m       int
		n       int
		hFences []int
		vFences []int
		expect  int
	}{
		{4, 3, []int{2, 3}, []int{2}, 4},
		{6, 7, []int{2}, []int{4}, -1},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.expect, maximizeSquareArea(test.m, test.n, test.hFences, test.vFences), index)
	}
}
